src/mame/jaleco/ms1_gatearray.cpp

```diff
@@ -0,0 +1,245 @@
+// license:BSD-3-Clause
+// copyright-holders:Angelo Salese, David Haywood
+
+#include "emu.h"
+#include "ms1_gatearray.h"
+
+//#define VERBOSE 1
+#include "logmacro.h"
+
+DEFINE_DEVICE_TYPE(MEGASYS1_GATEARRAY_D65006, megasys1_gatearray_d65006_device, "ms1_d65005", "Mega System 1 Gate Array D65005")
+DEFINE_DEVICE_TYPE(MEGASYS1_GATEARRAY_GS88000, megasys1_gatearray_gs88000_device, "ms1_gs88000", "Mega System 1 Gate Array GS-88000")
+DEFINE_DEVICE_TYPE(MEGASYS1_GATEARRAY_UNKARRAY, megasys1_gatearray_unkarray_device, "ms1_unkarray", "Mega System 1 Gate Array UNKNOWN")
+
+megasys1_gatearray_device::megasys1_gatearray_device(
+		machine_config const &mconfig,
+		device_type type,
+		char const *tag,
+		device_t *owner,
+		u32 clock)
+	: device_t(mconfig, type, tag, owner, clock)
+	, m_cpu(*this, finder_base::DUMMY_TAG)
+{
+	m_has_decoded = false;
+}
+
+
+void megasys1_gatearray_device::device_start()
+{
+	save_item(NAME(m_gatearray_hs));
+	save_item(NAME(m_gatearray_hs_ram));
+
+	install_gatearray_overlay();
+}
+
+void megasys1_gatearray_device::device_reset()
+{
+	if (!m_has_decoded)
+	{
+		rom_decode();
+		m_has_decoded = true;
+		m_cpu->reset();
+	}
+
+	m_gatearray_hs = 0;
+}
+
+
+void megasys1_gatearray_d65006_device::rom_decode()
+{
+	u16 *RAM = (u16 *)m_rom;
+	int size = m_romsize;
+	if (size > 0x40000) size = 0x40000;
+
+	for (int i = 0 ; i < size/2 ; i++)
+	{
+		u16 y;
+		u16 x = RAM[i];
+
+// [0] def0 189a bc56 7234
+// [1] fdb9 7531 eca8 6420
+// [2] 0123 4567 ba98 fedc
+#define BITSWAP_0   bitswap<16>(x,0xd,0xe,0xf,0x0,0x1,0x8,0x9,0xa,0xb,0xc,0x5,0x6,0x7,0x2,0x3,0x4)
+#define BITSWAP_1   bitswap<16>(x,0xf,0xd,0xb,0x9,0x7,0x5,0x3,0x1,0xe,0xc,0xa,0x8,0x6,0x4,0x2,0x0)
+#define BITSWAP_2   bitswap<16>(x,0x0,0x1,0x2,0x3,0x4,0x5,0x6,0x7,0xb,0xa,0x9,0x8,0xf,0xe,0xd,0xc)
```

Comment:
Please use lambdas for these, something along the lines of:
```c++
	auto const BITSWAP_0 = [x] () { return bitswap<16>(x,0xd,0xe,0xf,0x0,0x1,0x8,0x9,0xa,0xb,0xc,0x5,0x6,0x7,0x2,0x3,0x4); };
```
etc. to avoid the need for preprocessor expansion.